include/irr/asset/CTexturePacker.h

```diff
@@ -4,35 +4,96 @@
 #include "irr/asset/format/EFormat.h"
 #include "irr/asset/ICPUImage.h"
 #include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+#include "irr/core/math/morton.h"
 
 namespace irr {
 namespace asset
 {
 
+//TODO move this to different file
+template<typename addr_type>
+class CSquareAddressAllocator
+{
+protected:
+    using lin_addr_alctr_t = core::GeneralpurposeAddressAllocator<addr_type>;
+
+    const uint32_t ADDR_LAYER_SHIFT;
+    const uint32_t ADDR_MORTON_MASK;
+
+    core::smart_refctd_dynamic_array<uint8_t> m_alctrReservedSpace;
+    lin_addr_alctr_t m_addrAlctr;
+
+public:
+    CSquareAllocator(uint32_t _addrBitCntPerDim, uint32_t _squareSz) : 
+        ADDR_LAYER_SHIFT(_addrBitCntPerDim*2u),
+        ADDR_MORTON_MASK((1u<<ADDR_LAYER_SHIFT)-1u),
+        m_alctrReservedSpace(core::make_refctd_dynamic_array<decltype(m_alctrReservedSpace)>(
+            lin_addr_alctr_t::reserved_size(_squareSz*_squareSz, _squareSz*_squareSz, 1ull))
+        ),
+        m_addrAlctr(reinterpret_cast<void*>(m_alctrReservedSpace->data()), 0u, 0u, _squareSz*_squareSz, _squareSz*_squareSz, 1u)
+    {}
+
+    inline addr_type alloc_addr(size_t _subsquareSz)
+    {
+        const size_t cnt = _subsquareSz*_subsquareSz;
+        return m_addrAlctr.alloc_addr(cnt*cnt, cnt*cnt);
+    }
+
+    inline addr_type unpackAddress_x(addr_type _addr) const
+    {
+        return core::morton2d_decode_x(_addr&ADDR_MORTON_MASK);
+    }
+    inline addr_type unpackAddress_y(addr_type _addr) const
+    {
+        return core::morton2d_decode_y(_addr&ADDR_MORTON_MASK);
+    }
+    inline addr_type unpackAddress_layer(addr_type _addr) const
+    {
+        return (_addr>>ADDR_LAYER_SHIFT);
+    }
+};
+
 class CTexturePacker
 {
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t TILE_PADDING = 8u;
```

Comment:
you want this to be given to the `CTexturePacker` constructor, since depending on the anisotropic filter you want to use the packed textures with, the padding can be between 0 and 9 (half of max anisotropy, +1 if using bilinear filter)